# Database Final Project NOSQL Part

## What we have done:
## We have finished the database final project for SQL part, then we will convert them into NOSQL part

## 1. Convert data into NOSQL database Mongodb
### We have converted previous 11 tables into Mongodb, now we convert 4 new tables into Mongodb

In [76]:
%matplotlib inline
import pymongo
from pymongo import MongoClient
from pymongo.errors import ConnectionFailure
from bson import json_util, ObjectId
import pandas as pd
from pandas import DataFrame
from pandas.io.json import json_normalize
import numpy as np
import requests
import json, os
import configparser
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats
import seaborn as sns
import warnings
import random
import pprint
from datetime import datetime
random.seed(datetime.now())
warnings.filterwarnings('ignore')

# Make plots larger
plt.rcParams['figure.figsize'] = (10, 6)

In [77]:
import csv
import pandas as pd
import pymysql
import tweepy
import json
from dateutil import parser
import pandas as pd
import re
import pymysql
import sqlalchemy

In [78]:
def sql_to_df(sql_query):
    df=pd.read_sql(sql_query,db)
    return df

In [83]:
db = pymysql.connect(host = '127.0.0.1', port = 3306, user = 'root', db = 'assignment2',charset='utf8mb4')
cursor = db.cursor()
client = MongoClient()
mdb=client['analytics']

In [84]:
a = '''SELECT * FROM Youtubetable;'''
b= sql_to_df(a)
bt = b.T
btd = bt.to_dict()

a =[]
for i in range(len(btd)):
    a.append(btd[i])

result = client.analytics.Youtubetable.insert_many(a)

In [87]:
a = '''SELECT * FROM tag_synonmys_table;'''
b= sql_to_df(a)
bt = b.T
btd = bt.to_dict()

a =[]
for i in range(len(btd)):
    a.append(btd[i])

result = client.analytics.tag_synonmys_table.insert_many(a)

In [88]:
a = '''SELECT * FROM mis_spelled_table;'''
b= sql_to_df(a)
bt = b.T
btd = bt.to_dict()

a =[]
for i in range(len(btd)):
    a.append(btd[i])

result = client.analytics.mis_spelled_table.insert_many(a)

In [89]:
a = '''SELECT * FROM semantic_information_table;'''
b= sql_to_df(a)
bt = b.T
btd = bt.to_dict()

a =[]
for i in range(len(btd)):
    a.append(btd[i])

result = client.analytics.semantic_information_table.insert_many(a)

## 2. Answer 12 questions for project in NOSQL

In [4]:
def process(l):
    delnum = []
    for i in range(len(l)):
        if l[i] == '':
            delnum.append(i)
            continue
        if l[i] in l[:i]:
            delnum.append(i)
            continue
    for i in reversed(delnum):
        l.pop(i)
    return l 

# Problem i

In [93]:
bp = '''SELECT text, hashtags FROM assignment2.tweettable
where screen_name = "ioTSC";'''
result=mdb.tweettable.find({'screen_name':'ioTSC'}, {'text':1,'hashtags':1})
res = []
for i in result:
    res.append(i)
pd.DataFrame(process(res), columns = ['text', 'hashtags'])

text  \
0   RT @ioTSC: @idamaria_pacino @ImPAN0 @AFP @CNBC...   
1   RT @idamaria_pacino: @ioTSC @lux_chiara @idama...   
2   RT @ioTSC: @lux_chiara @idamariapan1 @idampan ...   
3   RT @ioTSC: @lux_chiara @idamariapan1 @idampan ...   
4   RT @ioTSC: @lux_chiara @idamariapan1 @idampan ...   
5   RT @idampan: @idamaria_pacino @ioTSC @impxyz @...   
6   RT @idampan: @idamaria_pacino @ioTSC @impxyz @...   
7   RT @idampan: @idamaria_pacino @ioTSC @impxyz @...   
8   RT @RobertDowneyJr: Happy Valentine’s Day to a...   
9   RT @idamaria_pacino: @RobertDowneyJr @idampan ...   
10  RT @RobertDowneyJr: Happy Labor Day! May every...   
11  RT @idamaria_pacino: @ImPAN0 @idampan @RobertD...   
12  RT @idamaria_pacino: @ioTSC @idampan @idamaria...   
13  RT @idamaria_pacino: @ImPAN0 @idampan @idamari...   
14  RT @ioTSC: @ImPAN0 @engadget @mskerryd @Robert...   
15  RT @idamaria_pacino: @idampan @richardbranson ...   
16  RT @ioTSC: @idampan @FBI @Reuters @JimmyKimmel...   
17  RT @ImPAN0: @ioTSC @lux_chiara @Marvel @idampa...   
18  RT @ioTSC: @impxyz @HeleenWolf @GiulsVezzoli @...   
19  RT @ImPAN0: @ioTSC @Disney #idampan #idaZERO #...   

                                             hashtags  
0                                                      
1                                                      
2                                                      
3                                                      
4                                                      
5                                   idampan photoimp   
6                                      idampan RDJ R   
7                                   idampan idaCohen   
8                                                      
9                     idampan idaTELL ida0 idaCarrey   
10                                                     
11                   idampan idaSMA idaTELL idaCohen   
12                                                     
13                                                     
14                                           idampan   
15                                                     
16                           idampan idaTell idaBond   
17                idampan idaBond idaFBI FBI idaKhan   
18                                                     
19  idampan idaZERO ida0 idaSMA idaXYZ photoimp BL...

### Problem i answer: the result is the hashtags that related to the user called ioTSC, which can be represent as what people are saying about him/her.

# Problem ii

In [94]:
bp = '''SELECT text, hashtags FROM assignment2.tweettable
where screen_name = "ioTSC";'''
result=mdb.tweettable.find({"tweet_id" : 976229529063600128}, {'retweetcounts':1,'favorites':1})
res = []
for i in result:
    res.append(i)
pd.DataFrame(process(res), columns = ['retweetcounts', 'favorites'])

retweetcounts  favorites
0            623        943

### Problem ii answer: the result is the retweet count and favourite count of the tweet whose tweet_id is 976229529063600128 which is a tweet by ioTSC. By seeing the retweet count and favourite count, we can know how many people saw this post roughly which is a measure to the question that how viral is the post.

# Problem iii

In [95]:
bp = '''SELECT text, hashtags FROM assignment2.tweettable
where screen_name = "ioTSC";'''
result=mdb.tweettable.find({"tweet_id" : 976229529063600128}, {'followers':1,'friends':1,'favourites':1,'retweetcounts':1,'favorites':1})
res = []
for i in result:
    res.append(i)
pd.DataFrame(process(res), columns = ['followers','friends','favourites','retweetcounts','favorites'])

followers  friends  favourites  retweetcounts  favorites
0        814     3586       33218            623        943

### Problem iii answer: we also count the followers, friends, favourites of the user-ioTSC. The influence of a post can be measure by how many people can see the post, which include the people who didn't like or retweet the post. And the followers can give us a rough knowledge how influence is the post

# Problem iv

    We take the following tweet_id as my tweet's id

In [96]:
result=mdb.tweettable.find({"tweet_id" : 976229241414176768},{'text':1,'hashtags':1})
res = []
for i in result:
    res.append(i)
pd.DataFrame(process(res), columns = ['text','hashtags'])

text  \
0  4 more weeks for @Avengers #infinitywar comes ...   

                      hashtags  
0  infinitywar CaptainAmerica

In [97]:
from difflib import SequenceMatcher
def similar(a, b):
    return SequenceMatcher(None, a, b).ratio()
result=mdb.tweettable.find({"hashtags" : {'$regex':'nfinity'}},{'text':1,'hashtags':1})
res2=[]
for i in result:
    res2.append(i)

df=pd.DataFrame(process(res2), columns = ['text','hashtags'])
a=[]
for i,j in enumerate(df['hashtags']):
    if similar(j,'infinitywar')>0.75:
        a.append(df['text'][i])
pd.DataFrame(list(set(a)))

0
0   RT @letitiawright: squad goals ?? x @Avengers ...
1   RT @TheWalkingDead: Marvel: "#InfinityWar is t...
2   RT @THR: New @Avengers: #InfinityWar trailer n...
3   RT @marvelmex: Avengers: #InfinityWar ha super...
4   Just came home from the cinema where the #Infi...
5   RT @tomhupdates: Tom Holland as Peter Parker a...
6   RT @Avengers: Watch the new Marvel Studios’ "A...
7   RT @Marvel: Watch the new Marvel Studios’ @Ave...
8   RT @chadwickboseman: New @Avengers: #InfinityW...
9   When it looks like the Avengers will make thei...
10  RT @Marvel: Here's a look at six new @EmpireMa...
11  RT @Dynamitejimi: The order to watch before #I...
12  RT @Avengers: Here's a look at six new @Empire...
13               #InfinityWar https://t.co/VGasw45j11
14  RT @Fandango: Be the first to see @Avengers: #...

### Problem iv answer: we take the tweet whose tweet_id is 976229241414176768 as my post. the result is the posts that liek my post.

# Problem v

In [98]:
result=mdb.tweettable.find({"hashtags" : {'$regex':'nfinity'}},{'screen_name':1,'hashtags':1})
res=[]
for i in result:
    res.append(i)
df=pd.DataFrame(process(res), columns = ['screen_name','hashtags'])
a=[]
for i,j in enumerate(df['hashtags']):
    if similar(j,'infinitywar')>0.75:
        a.append(df['screen_name'][i])
pd.DataFrame(list(set(a)))

0
0          J_Bologna
1         sjhajdzjd2
2          maddcocoa
3           ftniazhr
4       muqirfanrhzd
5              exz6y
6         Gojira007X
7        DannyBeegie
8         Smithryan8
9            covos__
10      hussein66783
11     Bass_Not_Bass
12     yoder_madison
13         asgard_GB
14        Cold_Sword
15          jusroyce
16         Art__Kino
17     Kaname_Lagann
18          crdude45
19     dk_inthehouse
20         Bievixnia
21     HeCallsMeKiwi
22    AthleticQuails
23     PrincessChez_
24         ritari_ta
25     Judai__gotcha
26      YourboyQuan_
27   AngelMo06304753
28          hydeist7
29         aboutnici
..               ...
78      daredevil_eg
79         LitassGuz
80    llewellyn_onyx
81   ofrancethemovie
82           ktaegod
83   JeremyPalencia8
84         tmrspidey
85    WritingHipster
86    muazzimuhammad
87       Lothiriel_1
88    imsexyandimNOE
89     cesiah_madrid
90         aMieBeckz
91       AlexdAlbers
92            bwoceo
93      RafaellXXX69
94     wigettamydads
95      MBCognitionz
96    RebelInTheDark
97     shawnjgilbert
98        Johnnyvle_
99          Jessmjr7
100      BarneyBrian
101     _QueenofSin_
102          n0namae
103       wandadesti
104       Abbah_Teee
105       TylerRoby1
106    LupeBolanos15
107          mrmozez

[108 rows x 1 columns]

### Problem v answer: We choose the users who post tweets contain hashtags contain 'nifinity', so we can get hashtags related to 'Avengers: Infinity War'. And the result is the users who post similar posts that like mine.

# Problem vi

In [99]:
result=mdb.tweettable.find({"hashtags" : {'$regex':'nfinity'}},{'screen_name':1,'hashtags':1,'followers':1})
res=[]
for i in result:
    res.append(i)
df=pd.DataFrame(process(res), columns = ['screen_name','hashtags','followers'])
a=[]
for i,j in enumerate(df['hashtags']):
    if similar(j,'infinitywar')>0.75:
        if df['followers'][i]>10000:
            a.append(df['screen_name'][i])
pd.DataFrame(list(set(a)))

0
0  FootSniffah
1      ktaegod

### Problem vi answer: From the result in problem v, we get users whose followers is larger than 1000. Then the result is the users I should following.

# Problem vii

In [100]:
resultup = mdb.tags.find({'trend':'up'},{'id':1,'tags':1,'trend':1, '_id':0})
pd.DataFrame(([i for i in resultup]))

id            tags trend
0  tt2873282      RedSparrow    up
1  nm0262635      ChrisEvans    up
2  nm0000123   georgeclooney    up
3  nm0005351    RyanReynolds    up
4  nm0000040  StanleyKubrick    up
5         04           Oscar    up

In [101]:
resultdown = mdb.tags.find({'trend':'down'},{'id':1,'tags':1,'trend':1,'_id':0})
pd.DataFrame(([i for i in resultdown]))

id                 tags trend
0   tt5580390      TheShapeofWater  down
1   tt1825683         BlackPanther  down
2          00       TheWalkingDead  down
3   tt5027774      ThreeBillboards  down
4   tt4154756  AvengersInfinityWar  down
5   tt5726616         CallMeByYour  down
6   tt2283362              Jumanji  down
7   tt2798920         Annihilation  down
8   tt1365519           TombRaider  down
9   nm0000375       RobertDowneyJr  down
10  nm2225369     jenniferlawrence  down
11  nm0000136           johnnydepp  down
12  nm0001467            JaredLeto  down
13  nm0331516          ryangosling  down
14  nm0000233     QuentinTarantino  down
15         01       CharlieChaplin  down
16         03      AlfredHitchcock  down
17  nm0000217       MartinScorsese  down
18  nm0000229      StevenSpielberg  down
19         05            Hollywood  down

In [102]:
resultnochange = mdb.tags.find({'trend':'no change'},{'id':1,'tags':1,'trend':1,'_id':0})
pd.DataFrame(([i for i in resultnochange]))

id               tags      trend
0  nm1475594     channing tatum  no change
1  nm0424060  ScarlettJohansson  no change
2  nm0000697        BillyWilder  no change
3  nm0000041      AkiraKurosawa  no change
4         02      HayaoMiyazaki  no change
5  nm0634240   ChristopherNolan  no change

### Problem vii answer: the result shows whether the tag is trending or not

# Problem viii

In [103]:
result=mdb.title.aggregate([{
    '$lookup':{
        'from':'tweettable',
        'localField':'tconst',
        'foreignField':'id',
        'as':'c'}
},
{
    '$match':{'$and':[{'c':{'$ne':[]}},
             {'primaryTitle':'Avengers: Infinity War'},
                     {}]}
}
])
res = []
for i in result:
    for j in i['c']:
        res.append((j['hashtags']))
from collections import Counter
count=Counter(list(filter(None, res)))
pd.DataFrame(count.most_common(),columns=['hashtag','count']).head(3)
#print(Counter(res))

hashtag  count
0           InfinityWar     124
1  Avengers InfinityWar      22
2   AvengersInfinityWar      21

### Problem viii answer: we get top 3 most used hashtags that related to 'Avengers: Infinity War'. the result is the hashtags that I should add to my post.

# Problem ix

In [104]:
result=mdb.tweettable.find({"hashtags" : {'$regex':'nfinity'}},{'screen_name':1,'hashtags':1})
res=[]
for i in result:
    res.append(i)
df=pd.DataFrame(process(res), columns = ['screen_name','hashtags'])
a=[]
for i,j in enumerate(df['hashtags']):
    if similar(j,'infinitywar')>0.75:
        a.append(df['screen_name'][i])
pd.DataFrame(list(set(a)))

0
0          J_Bologna
1         sjhajdzjd2
2          maddcocoa
3           ftniazhr
4       muqirfanrhzd
5              exz6y
6         Gojira007X
7        DannyBeegie
8         Smithryan8
9            covos__
10      hussein66783
11     Bass_Not_Bass
12     yoder_madison
13         asgard_GB
14        Cold_Sword
15          jusroyce
16         Art__Kino
17     Kaname_Lagann
18          crdude45
19     dk_inthehouse
20         Bievixnia
21     HeCallsMeKiwi
22    AthleticQuails
23     PrincessChez_
24         ritari_ta
25     Judai__gotcha
26      YourboyQuan_
27   AngelMo06304753
28          hydeist7
29         aboutnici
..               ...
78      daredevil_eg
79         LitassGuz
80    llewellyn_onyx
81   ofrancethemovie
82           ktaegod
83   JeremyPalencia8
84         tmrspidey
85    WritingHipster
86    muazzimuhammad
87       Lothiriel_1
88    imsexyandimNOE
89     cesiah_madrid
90         aMieBeckz
91       AlexdAlbers
92            bwoceo
93      RafaellXXX69
94     wigettamydads
95      MBCognitionz
96    RebelInTheDark
97     shawnjgilbert
98        Johnnyvle_
99          Jessmjr7
100      BarneyBrian
101     _QueenofSin_
102          n0namae
103       wandadesti
104       Abbah_Teee
105       TylerRoby1
106    LupeBolanos15
107          mrmozez

[108 rows x 1 columns]

### Problem ix answer: This question is similar to the question vi. Thus, if someone in the result of problem v follow me, I should follower him

# Problem x

In [105]:
import datetime
import time
def get_sec(time_str):
    h, m, s = time_str.split(':')
    return int(h) * 3600 + int(m) * 60 + int(s)
result=mdb.tweettable.find({'followers':{'$gt':1000}},{'created_at':1})
df=pd.DataFrame(([i['created_at'] for i in result]))
j=0
for i in df[0]:
    x=datetime.datetime.strftime(i,'%H:%M:%S')
    j=j+(get_sec(x))
str(datetime.timedelta(seconds=j/len(df)))

'22:50:56.036496'

### Problem x answer: We calculated the average time of these tweets. And the result shows the best time to post. Because the time is UTC, to adjust it into EST, we should substract 5.

# Problem xi

In [106]:
client = MongoClient()
mdb=client['analytics']
total = mdb.tweettable.aggregate([
    {'$group': {'_id': "$text", 'num_tutorial': {'$sum':1}}},
    {'$sort': {"num_tutorial":-1}},
    {'$match':{'num_tutorial': {'$gt': 1} }},
    {'$count': 'total'}
])
total = pd.DataFrame(([i for i in total]))


u = mdb.tweettable.aggregate([
    {'$group': {'_id': "$text", 'num_tutorial': {'$sum':1}}},
    {'$sort': {"num_tutorial":-1}},
   
    {'$match':{'$and':[{'num_tutorial': {'$gt': 1}}, {'_id': {'$regex': 'https://'}}]}},
    {'$count': 'url'}
    
])
u = pd.DataFrame(([i for i in u]))

rate = float(u.iloc[0][0])/total.iloc[0][0]
rate

0.5374149659863946

### Problem xi answer: because the result number is 0.5374149659863946, which is larger than 0.5, thus we should add url to my post.

# Problem xii

In [107]:
result=mdb.tweettable.find({'tweet_id':976229529063600128},{'retweetcounts':1,'favorites':1,'followers':1})
r=pd.DataFrame(([i for i in result]))
((r['retweetcounts']+r['favorites'])/r['followers'])[0]

1.9238329238329239

### Problem xii answer: This result is calculated by the amount of people this post can infulence divided by the followers of the user. We get the number 1.9, which means that the post not only reach to the friends of the user but also the friends of his/her friends.